File: produto_escalar/Scala/ProdutoEscalar.scala
package br.com.ciandt.dojo.produtoEscalar;
object ProdutoEscalar {

  def main(args: Array[String]) {
    val a = List(1,2)
    val b = List(1,2,3)
    
    println(calcula(a,b))
    
  }
  
 
  def calcula(a: List[Int], b: List[Int]): String = {
    
    if (a.length != b.length) throw new IllegalArgumentException("Tamanhos diferentes")
    
    calculoAngularidade(calculaProduto(a, b))
  }
  
  private def calculaProduto(a: List[Int], b: List[Int]): Int = {
    (a,b).zipped.map(_*_).fold(0)(_+_)
  }
  
  private def calculoAngularidade(produto : Int) : String = {
    
    if (produto == 0) return "Reto"
   
    produto / produto.abs match {
      case -1 => "Obtuso"
      case 1  => "Agudo"
    }
  }

}

File: produto_escalar/php/ProdutoEscalar.php
<?php
/**
 * Coding Dojo Ci&T
 * 
 * @author Leandro Leonardo Isotton (lisotton)
 * @author Alberto Viana (albertov)
 * @author Celso Lopes de Andrade (clopes)
 * 
 * Problema:Desenvolva um programa que, dado dois vetores de dimensão n, retorne o produto escalar entre eles.
 */

class ProdutoEscalar {
  
  /**
   * Calcula o produto escalar de duas listas
   * @param array $a
   * @param array $b
   * @return array
   * @throws InvalidArgumentException 
   */
  public function calculate($a, $b) {
    if (!is_array($a) || !is_array($b)) {
      throw new InvalidArgumentException('Wrong Arguments, needs to be arrays');
    }
    
    if (!count($a) || !count($b)) {
      throw new InvalidArgumentException('Wrong Arguments, size of array needs to be more than one');
    }
    
    if (count($a) !== count($b)) {
      throw new InvalidArgumentException('Wrong Arguments, size of arrays needs to be the same');
    }
    
    $product = 0;
    foreach ($a as $key => $value) {
      $product += $value * $b[$key];
    }
    return array($product, $this->get_angle($product));
  }
  
  /**
   * Retorna o angulo do produto
   * @param int $product
   * @return string 
   */
  public static function get_angle($product) {
    if ($product < 0) {
      return 'obtuso';
    }
    elseif ($product > 0) {
      return 'agudo';
    }
    else {
      return 'reto';
    }
  }
  
}

?>
